napari/layers/utils/transform_utils.py

```diff
@@ -0,0 +1,198 @@
+    """
+    if np.isscalar(rotate):
+        # If a scalar is passed assume it is a single rotate angle
+        # for a 2D rotate
+        if degrees:
+            theta = np.deg2rad(rotate)
+        else:
+            theta = rotate
+        rotate_mat = np.array(
+            [[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]]
+        )
+        # convert to numpy coordinates
+        rotate_mat = rotate_mat[::-1, ::-1]
+    elif np.array(rotate).ndim == 1 and len(rotate) == 3:
+        # If a 3-tuple is passed assume it is three rotation angles for
+        # a roll, pitch, and yaw for a 3D rotation. For more details see
+        # https://en.wikipedia.org/wiki/Rotation_matrix#General_rotations
+        if degrees:
+            alpha = np.deg2rad(rotate[0])
+            beta = np.deg2rad(rotate[1])
+            gamma = np.deg2rad(rotate[2])
+        else:
+            alpha = rotate[0]
+            beta = rotate[1]
+            gamma = rotate[2]
+        R_alpha = np.array(
+            [
+                [np.cos(alpha), np.sin(alpha), 0],
+                [-np.sin(alpha), np.cos(alpha), 0],
+                [0, 0, 1],
+            ]
+        )
+        R_beta = np.array(
+            [
+                [np.cos(beta), 0, np.sin(beta)],
+                [0, 1, 0],
+                [-np.sin(beta), 0, np.cos(beta)],
+            ]
+        )
+        R_gamma = np.array(
+            [
+                [1, 0, 0],
+                [0, np.cos(gamma), -np.sin(gamma)],
+                [0, np.sin(gamma), np.cos(gamma)],
+            ]
+        )
+        rotate_mat = R_alpha @ R_beta @ R_gamma
+        # convert to numpy coordinates
+        rotate_mat = rotate_mat[::-1, ::-1]
+    else:
+        # Otherwise assume a full nD rotate matrix has been passed
+        rotate_mat = np.array(rotate)
+
+    # Convert a scale vector to an nD diagonal matrix
+    scale_mat = np.diag(scale)
+
+    # Check if an upper-triangular representation of shear or
+    # a full nD shear matrix has been passed
+    if np.array(shear).ndim == 1:
+        shear_mat = expand_upper_triangular(shear)
+    else:
+        shear_mat = np.array(shear)
+
+    # Check the dimensionality of the transforms and pad as needed
+    n_scale = scale_mat.shape[0]
+    n_rotate = rotate_mat.shape[0]
+    n_shear = shear_mat.shape[0]
+    ndim = max(n_scale, n_rotate, n_shear)
+
+    full_scale = embed_in_identity_matrix(scale_mat, ndim)
+    full_rotate = embed_in_identity_matrix(rotate_mat, ndim)
+    full_shear = embed_in_identity_matrix(shear_mat, ndim)
+    return full_rotate @ full_scale @ full_shear
+
+
+def expand_upper_triangular(vector):
+    """Expand a vector into an upper triangular matrix.
+
+    Decomposition is based on code from https://github.com/matthew-brett/transforms3d.
+    In particular, the `striu2mat` function in the `shears` module.
+    https://github.com/matthew-brett/transforms3d/blob/master/transforms3d/shears.py#L30.
```

Comment:
fwiw: I like that we are referencing external libraries to both give credit and ensure we can trace the ideas. One minor pick is that linking out to external files without a tagging commit can later become irrelevant and confusing if they are to update their codebase.

Reply:
good catch!! i'm tagging to latest release, so looks like "https://github.com/matthew-brett/transforms3d/blob/0.3.1/transforms3d/shears.py#L30-L77"